docs/design/adrs/2024-04-23-implement-prompflow-orchestration.md

```diff
@@ -0,0 +1,42 @@
+# Implement a PromptFlow Orchestration as an alternative for the chat backend
+
+* **Status:** proposed
+* **Proposer:** @superhindupur
+* **Date:** 2024-04-23
+* **Technical Story:** [Include a prompt flow integration](https://github.com/Azure-Samples/chat-with-your-data-solution-accelerator/issues/406)
+
+## Context and Problem Statement
+
+The chat functionality in this repo has an orchestraction mechanism for fetching the answer based on the user's question - the question is sent to the LLM, then based on whether it needs to be searched, it goes to the search service before the answer is finally reprocessed and sent back to the user on the chat UI.
+
+This orchestration between the chat backend, LLM and search service is done inside code and currently has two options for orchestration management - openai functions and langchain (controlled by the deployment bicep param `orchestrationStrategy`). In this ADR, we explore whether it is worth adding a third mechanism to orchestrate the chat backend functionality with [Azure Machine Learning PromptFlow](https://learn.microsoft.com/en-us/azure/machine-learning/prompt-flow/overview-what-is-prompt-flow?view=azureml-api-2).
+
+## Decision Drivers
+
+* Being able to demonstrate various orchestration options to customers
+* Limiting complexity in the repo
+* Ease of maintenance of the PromptFlow configuration
+
+## Considered Options
+
+* Adding the PromptFlow orchestration
+* Not adding the PromptFlow orchestration
+
+## Decision Outcome
+In the absence of concrete business / customer requirements, the outcome is to NOT add a PromptFlow orchestrator alternative for the chat backend.
+
+
+## Pros and Cons of the Options
+
+### Adding the PromptFlow orchestration
+* Good, because customers wanting to deploy the chat backend as a PromptFlow endpoint will be able to do so.
+* Good, because customers will have a low-code mechanism to experiment with the orchestration of the chat backend.
+* Bad, because building a PromptFlow orchestration that mimics the current chat backend requires significant effort - and it is unclear how much customer demand we have for it.
+* Bad, because the PromptFlow configuration can get out of date with the other orchestrators that are in code. This will cause confusion to customers and will lead them to open issues against the repo.
```

Comment:
This is the same for langchain vs openai functions vs semantic kernel. Do you think we should align on just one orchestrator as a favourite?

Reply:
Isn't semantic kernel a code-based orchestrator? If that's the case wouldn't the interfaces in code ensure that the functionality will remain common across it and other code-based orchestrators? If we don't have interfaces already we at least have the possibility of creating them and enforcing uniformity if they're in code.

With PF, that's not an option because the orchestration is in a dag file.

Reply:
> Isn't semantic kernel a code-based orchestrator?

Yea correct. There's only one common method which is:

```python
    @abstractmethod
    def orchestrate(
        self, user_message: str, chat_history: List[dict], **kwargs: dict
    ) -> dict:
        pass
```
https://github.com/Azure-Samples/chat-with-your-data-solution-accelerator/blob/3e71fe10c50a129a20c314ba015065acdc6d8379/code/backend/batch/utilities/orchestrator/OrchestratorBase.py#L27-L31

Each orchestrator is then free to do what it likes within that. 

Obviously it's extra effort, but we should ensure as reviewers all changes made and features added are applied to all orchestration types.

Reply:
Fair enough, took this into consideration while making the final decision.